File: 백준/Gold/12869. 뮤탈리스크/뮤탈리스크.js
const input = require('fs').readFileSync(process.platform === 'linux' ? '/dev/stdin' : './input.txt').toString().trim().split('\n');

const N = Number(input[0]);
const scv = input[1].split(' ').map(Number);
while(scv.length !== 3) scv.push(0);

const attack = [[9, 3, 1], [9, 1, 3], [3, 9, 1], [3, 1, 9], [1, 9, 3], [1, 3, 9]];
const visited = Array.from({length: 64}, () => Array.from({length: 64}, () => Array(64).fill(0)));

function answer(){
    const queue = [];

    queue.push(scv);
    visited[scv[0]][scv[1]][scv[2]] = 1;

    while(queue.length){
        const [a, b, c] = queue.shift();

        for (const atk of attack){
            const na = Math.max(a - atk[0], 0);
            const nb = Math.max(b - atk[1], 0);
            const nc = Math.max(c - atk[2], 0);

            if(visited[na][nb][nc]) continue;
            queue.push([na, nb, nc]);
            visited[na][nb][nc] = visited[a][b][c] + 1;        
        }
    }
    console.log(visited[0][0][0] - 1);
}

answer();

File: 백준/Gold/12869. 뮤탈리스크/뮤탈리스크.cc
#include <bits/stdc++.h>
using namespace std;

struct A {
  int a, b, c;
};

int attacks[6][3] = {{9, 3, 1}, {9, 1, 3}, {3, 9, 1},
                     {3, 1, 9}, {1, 9, 3}, {1, 3, 9}};

int n, scv[3];
int visited[62][62][62];
queue<A> q;

int main() {
  cin >> n;
  for (int i = 0; i < n; i++) {
    cin >> scv[i];
  }

  visited[scv[0]][scv[1]][scv[2]] = 1;
  q.push({scv[0], scv[1], scv[2]});

  while (q.size()) {
    int a = q.front().a;
    int b = q.front().b;
    int c = q.front().c;
    q.pop();
    if (a == 0 && b == 0 && c == 0) continue;

    for (int i = 0; i < 6; i++) {
      int na = max(0, a - attacks[i][0]);
      int nb = max(0, b - attacks[i][1]);
      int nc = max(0, c - attacks[i][2]);

      if (visited[na][nb][nc]) continue;

      visited[na][nb][nc] = visited[a][b][c] + 1;
      q.push({na, nb, nc});
    }
  }

  cout << visited[0][0][0] - 1 << '\n';

  return 0;
}